Add promo-blocking UI to React

Vinnl opened the pull request:
# New feature description

This implements the promotional email blocking feature in React. It's implemented as a [slider](https://react-spectrum.adobe.com/react-aria/useSlider.html), so it can be focused with the keyboard, and then switch states using the arrow keys. Since sliders need to control a numeric value (which might get exposed through screen readers?), I used 0 (no blocking), 50 (only block promotional emails), and 100 (block everything).

# Screenshot (if applicable)

![image](https://user-images.githubusercontent.com/4251/160109532-99a54ea9-a258-4dc4-823b-a4a2d3bc0671.png)

![image](https://user-images.githubusercontent.com/4251/160109557-518dfc3b-b164-415c-96af-541eb3db88f0.png)

# How to test

Visit https://deploy-preview-1688--fx-relay-demo.netlify.app/?mockId=some, you should see the slider in the alias cards there.

# Checklist

- [x] l10n dependencies have been merged, if any.
- [x] All acceptance criteria are met.
- [x] I've added or updated relevant docs in the docs/ directory.
- [x] All UI revisions follow the [coding standards](https://github.com/mozilla/fx-private-relay/blob/main/docs/coding-standards.md), and use Protocol tokens where applicable (see `/static/scss/libs/protocol/css/includes/tokens/dist/index.scss`).
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

Vinnl commented:
Rebased to resolve conflicts.

Vinnl commented on frontend/src/components/dashboard/aliases/Alias.tsx:
```diff
@@ -104,12 +107,66 @@ export const Alias = (props: Props) => {
+  // We have the <BlockLevelSlider> for Premium users, so don't show the toggle
+  // for them:
+  const toggleButton = props.profile.has_premium ? (
+    // An empty space can take the cell in the grid layout that otherwise
```
Yes! :eagle: :eye: 9475d087

Vinnl commented on frontend/src/components/dashboard/aliases/Alias.tsx:
```diff
@@ -104,12 +107,66 @@ export const Alias = (props: Props) => {
+        {l10n.getString("profile-promo-email-blocking-label-promotionals")}
+      </b>
+    );
+  }
```
1de4a830

Vinnl commented on frontend/src/components/dashboard/aliases/BlockLevelSlider.tsx:
```diff
@@ -0,0 +1,270 @@
+              ) === "none"
+                ? styles["is-active"]
+                : ""
+            }`}
```
Oof yeah, I think the camelCase to snake-case conversion did not help there either. Clarified in fa150642.

Vinnl commented on frontend/src/components/dashboard/aliases/BlockLevelSlider.tsx:
```diff
@@ -0,0 +1,270 @@
+              ) === "promotional"
+                ? styles["is-active"]
+                : ""
+            }`}
```
fa150642

Vinnl commented on frontend/src/components/dashboard/aliases/BlockLevelSlider.tsx:
```diff
@@ -0,0 +1,270 @@
+              ) === "all"
+                ? styles["is-active"]
+                : ""
+            }`}
```
fa150642

Vinnl commented on frontend/src/components/dashboard/aliases/BlockLevelSlider.tsx:
```diff
@@ -0,0 +1,270 @@
+        label: label,
+      });
+      return props.onChange(blockLevel);
+    },
```
I'm presuming you're primarily referring to the nested ternary (should've been a red flag when writing it)? 9fc4ca52

Or would you prefer I also assign the `onChange` function as a whole to a helper variable?

groovecoder replied:
double-ternary for sure. there's 3 lines/statements in this inline function definition, which feels kinda close to the a max number of statements we ever want to have in an inline function definition?

Vinnl replied:
Yeah, I wouldn't put a hard-and-fast number on it, but this does feel just about OK to me.

Vinnl commented on frontend/src/components/dashboard/aliases/BlockLevelSlider.tsx:
```diff
@@ -0,0 +1,270 @@
+          props.sliderState.isThumbDragging(onlyThumbIndex)
+            ? styles["is-dragging"]
+            : ""
+        }`}
```
2e9a7981

Vinnl commented:
> question: we have mock data/responses defined in both apiMocks/handlers.ts and apiMocks/mockData.ts - what's the difference? can we consolidate?

Some of the difference was legacy from before I split them out - those are consolidated in bec07e52. There is still _some_ mock data in `handlers.ts`, but that's data that's newly added to the mocks, i.e. data added through POST or PATCH (e.g. a new alias). Since those are dynamically generated (e.g. with the creation date set to now) and not part of the mock profiles by default, they're inline in the mock API handlers.

> question: I can't tell if we have a standard way to determine when we embed a component via a declared component vs. a variable 

We don't, although based on your reviews so far it feels like we'll be leaning towards components. I guess it's a pretty similar question to when to split something out into separate functions rather than calculating it inline - I've often gravitated towards keeping things inline to avoid the reader having to jump around and potentially lose context, but I'm sympathetic towards more concise functions (/components) as well.

maxxcrawford commented on frontend/src/components/dashboard/aliases/Alias.module.scss:
```diff
@@ -171,6 +171,32 @@ $toggleTransitionDuration: 200ms;
+  display: none;
+  align-items: center;
+
+  @media screen and #{$mq-md} {
```
Praise: Noticed that all the media queries in `/frontend` have `@media screen and` instead of just the `min-width` arguments. I dig the specificity.

Vinnl replied:
Heh, to be fair, that's VSCode autocompleting that for me when I type `@media` :P

maxxcrawford reviewed (approved):
Approving, but there's a few visual issues comparing the Figma, that did not get ported over from the Django implementation. The slider is fantastic!

**Missing: The UI is not visible on non-premium accounts.** 

It should be visible but disabled. _(FWIW, the Django implementation did not have this either, but that was bc React was coming)_

Figma: 
<img width="702" alt="image" src="https://user-images.githubusercontent.com/2692333/161633885-d5b50b7d-f256-4373-8126-542b420d0fc6.png">

**Missing: The selected state is missing from the text: (It should be darker/bold)** 
Figma: 
<img width="442" alt="image" src="https://user-images.githubusercontent.com/2692333/161634218-a381192f-2093-40d6-be41-9a925622fb46.png">

PR:
<img width="572" alt="image" src="https://user-images.githubusercontent.com/2692333/161634526-8843cad1-b111-4a4c-9009-dccfc35d115f.png">


**The illustration should be taller:** 

PR: 
<img width="484" alt="image" src="https://user-images.githubusercontent.com/2692333/161634391-4f1f6d3b-da16-4a02-8e24-7aad71650dba.png">

Figma: 
<img width="529" alt="image" src="https://user-images.githubusercontent.com/2692333/161634426-8cd363d3-91b3-4c94-8c67-62662754ca7e.png">

**Mobile: Missing/mismatched `max-width`'s:**

PR: 
<img width="597" alt="image" src="https://user-images.githubusercontent.com/2692333/161634800-406ed8ae-8fd9-4bf6-9da0-54a63df05680.png">

Figma: 
<img width="386" alt="image" src="https://user-images.githubusercontent.com/2692333/161634875-2fb82a91-97f3-483b-956d-0904ea881ada.png">

**Mobile: Forwarding label should be above the selector:** 

PR: 
<img width="405" alt="image" src="https://user-images.githubusercontent.com/2692333/161635360-12f9f279-1644-4604-aa9e-ca08543e2341.png">

Figma: 
<img width="388" alt="image" src="https://user-images.githubusercontent.com/2692333/161635006-95b6f6e5-c577-4708-93f0-9fac0e233b35.png">

**Mobile: Delete button should be in two-col layout with created date:** 

PR: 
<img width="423" alt="image" src="https://user-images.githubusercontent.com/2692333/161635208-9b36fb54-12a3-4e86-aa2d-3c437393b713.png">


Figma: 
<img width="402" alt="image" src="https://user-images.githubusercontent.com/2692333/161635124-286d145c-f509-4ce2-8251-8e01e096776a.png">

Vinnl commented on frontend/src/apiMocks/mockData.ts:
```diff
@@ -1,8 +1,37 @@
 
+// This is the same for all mock users, at this time:
+export const runtimeData: RuntimeData = {
+  FXA_ORIGIN: "https://example.com",
+  GOOGLE_ANALYTICS_ID: "UA-123456789-0",
+  PREMIUM_PRODUCT_ID: "prod_123456789",
+  PREMIUM_PLANS: {
+    country_code: "nl",
+    plan_country_lang_mapping: {
+      nl: {
+        nl: {
+          id: "price_1JmROfJNcmPzuWtR6od8OfDW",
+          price: "€0,99",
+        },
+      },
+    },
+    premium_countries: ["nl"],
+    premium_available_in_country: true,
+  },
```
Not necessarily a risk, but it could indeed be useful to also have a mock account for which Premium is not available, in case we want to test/demo UI in that situation. That's pretty easy to add using the same mechanism as used for other accounts, but let's cross that bridge when we get to it, because it's yet another mock to maintain.

Vinnl commented on frontend/src/components/dashboard/aliases/Alias.module.scss:
```diff
@@ -171,6 +171,32 @@ $toggleTransitionDuration: 200ms;
+      color: $color-light-gray-70;
+      border-color: $color-light-gray-40;
+    }
+    &.block-level-promotional-label {
```
2c681c60

Vinnl commented on frontend/src/components/dashboard/aliases/BlockLevelSlider.module.scss:
```diff
@@ -0,0 +1,209 @@
+            border-color: $color-purple-60;
+          }
+          &.is-dragging {
+            // background-color: dimgrey;
```
ac0df489

Vinnl commented on frontend/src/components/dashboard/aliases/BlockLevelSlider.module.scss:
```diff
@@ -0,0 +1,209 @@
+        }
+        &.track-stop-all {
+          right: calc(-1 * (var(--thumbDiameter) / 2));
+        }
```
Good point, I added comments for all calculations (and removed two calculations that were remnants from earlier experimentation and no longer relevant - another pro of adding comments) in e7a55349.

Vinnl commented on frontend/src/components/dashboard/aliases/BlockLevelSlider.module.scss:
```diff
@@ -0,0 +1,209 @@
+    .track {
+      $iconHeight: 50px;
+      $trackLineHeight: 4px;
+      --trackLineOffset: calc(
+        #{$iconHeight} + (var(--thumbDiameter) / 2) - (#{$trackLineHeight} / 2)
+      );
```
e7a55349

Vinnl commented:
Thanks for the very thorough review @maxxcrawford! As for

> Missing: The UI is not visible on non-premium accounts.

You'll be happy to know that I already have https://github.com/mozilla/fx-private-relay/pull/1673 or that :) I figured I'd split that out to lower the review load of this one.

> Missing: The selected state is missing from the text: (It should be darker/bold)

Oops, broke that in fa1506421779520c1b5042cf903e5a129c8806b7. Fixed in 28f8ba07.

> The illustration should be taller:

It looks like the illustrations in the Figma changed - the closed umbrella was no longer rotated, and they're bigger than they're on the current website. I inserted the new illustrations in d3375651.

> Mobile: Missing/mismatched max-width's:

Ah, it wasn't clear to me that that was a maximum width, but it does indeed look a bit weird on larger mobile screens. Added a max-width in d52288b1.

> Mobile: Forwarding label should be above the selector:
> Mobile: Delete button should be in two-col layout with created date:

These appear to be part of a bigger redesign of the alias cards on mobile (also showing the stats), so I think I'll leave that as part of that implementation (which I think is already planned as part of showing the reply count).

groovecoder reviewed (approved):
LGTM after Maxx's changes are in too. Let's merge and stage and QA!
